### PR TITLE
New pagination / iteration model.

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: src/client/pfs.go [hunk 1]
```diff
@@ -234,6 +235,25 @@ func (c APIClient) inspectCommit(repoName string, commitID string, blockState pf
 // `number` determines how many commits are returned.  If `number` is 0,
 // all commits that match the aforementioned criteria are returned.
 func (c APIClient) ListCommit(repoName string, to string, from string, number uint64) ([]*pfs.CommitInfo, error) {
```
Comment (brycemcanally) [inline src/client/pfs.go]: As a side note, it might be useful to have a commit range abstraction that holds the from, to, and number of commits.

File: src/server/pfs/server/driver.go [hunk 2]
```diff
@@ -1057,78 +1049,86 @@ func parseCommitID(commitID string) (string, int) {
...
-		if err != nil {
-			return nil, err
-		}
-		var commitID string
-		for number != 0 {
-			var commitInfo pfs.CommitInfo
-			ok, err := iterator.Next(&commitID, &commitInfo)
-			if err != nil {
-				return nil, err
+		if err := commits.ListF(col.Descend, ci, func(commitID string) error {
```
Comment (brycemcanally) [inline src/server/pfs/server/driver.go]: What happens here if number > 0 and less than the number of commits?

File: src/server/pfs/server/server_test.go [hunk 3]
```diff
@@ -2653,6 +2653,9 @@ func TestPutFileSplitDelete(t *testing.T) {
 
 	files, err := c.ListFile(repo, commit.ID, "line")
 	require.NoError(t, err)
+	for _, file := range files {
```
Comment (brycemcanally) [inline src/server/pfs/server/server_test.go]: Leftover debugging prints.

File: src/server/pkg/errutil/errutil.go [hunk 4]
```diff
@@ -1,9 +1,17 @@
...
 import (
+	"fmt"
 	"strings"
 )
 
+var (
+	// ErrBreak is an error used to break out of call back based iteration,
+	// should be swallowed by iteration functions and treated as successful
+	// iteration.
+	ErrBreak = fmt.Errorf("BREAK")
```
Comment (brycemcanally) [inline src/server/pkg/errutil/errutil.go]: This is just a note about changing this idiom to returning a break/done boolean, rather than returning a break within an error.

Comment (jdoliner) [reply src/server/pkg/errutil/errutil.go]: :+1: yeah, ErrBreak is pretty confusing and is likely to encourage buggy implementations of iteration methods.

File: src/server/pkg/collection/collection_test.go [hunk 5]
```diff
@@ -408,6 +398,79 @@ func TestTTLExtend(t *testing.T) {
 	require.True(t, actualTTL > TTL && actualTTL < LongerTTL, "actualTTL was %v", actualTTL)
 }
 
+func TestPagination(t *testing.T) {
```
Comment (brycemcanally) [inline src/server/pkg/collection/collection_test.go]: Should this be named TestIteration instead?

Comment (jdoliner) [reply src/server/pkg/collection/collection_test.go]: Yeah that probably makes more sense, or maybe just `TestList` I called it `TestPagination` because the backend logic that it's testing is the pagination of etcd requests. But probably makes more sense to name it based on the interface it's using, rather than the implementation it's using.

File: src/server/pkg/collection/collection.go [hunk 6]
```diff
@@ -493,77 +479,95 @@ func (c *readonlyCollection) GetBlock(key string, val proto.Message) error {
...
 
-func endKeyFromPrefix(prefix string) string {
-	// Lexigraphically increment the last character
-	return prefix[0:len(prefix)-1] + string(byte(prefix[len(prefix)-1])+1)
-}
-
-// ListPrefix returns lexigraphically sorted (not sorted by create time)
-// results and returns an iterator that is paginated
-func (c *readonlyCollection) ListPrefix(prefix string) (Iterator, error) {
+// ListPrefix returns keys (and values) that being with prefix, f will be
```
Comment (brycemcanally) [inline src/server/pkg/collection/collection.go]: This probably needs to be reworded.

File: src/server/pkg/collection/collection.go [hunk 7]
```diff
@@ -493,77 +479,95 @@ func (c *readonlyCollection) GetBlock(key string, val proto.Message) error {
...
+			}
+		}
+		var resp *etcd.GetResponse
+		var limit int64
+		for {
+			limit = atomic.LoadInt64(limitPtr)
+			var err error
+			resp, err = c.etcdClient.Get(c.ctx, prefix, append(opts, etcd.WithLimit(limit))...)
+			if err != nil {
+				if strings.Contains(err.Error(), "received message larger than max") {
```
Comment (brycemcanally) [inline src/server/pkg/collection/collection.go]: It might be better to check against an error code rather than the message because I have seen this message be different between grpc versions. It looks like the error code for this is ResourceExhausted.

File: src/client/pfs.go (hunk 1 above)
Comment (jdoliner) [reply src/client/pfs.go]: We do actually have a `CommitRange` proto message, but it's used in different places. It might make sense to group these together at some point.

File: src/server/pkg/errutil/errutil.go (hunk 4 above)
Comment (jdoliner) [reply src/server/pkg/errutil/errutil.go]: So having thought about this a little more I think I like what we have now better.
Having every return line in an iteration function have to specify that it doesn't want to break by returning `false` leads to a lot of clutter, and it prevents some particularly nice cases where we can pass GRPC `Send` methods directly to iteration functions since they have the right signature.
I also did some more reading of the standard library [function](https://golang.org/pkg/path/filepath/#Walk) this approach is based on, and they do a similar thing where they have a special purpose error that the callback can return, to control the flow of iteration.

Comment (brycemcanally) [reply src/server/pkg/errutil/errutil.go]: Okay, I am convinced. Although it is still kind of weird, the benefit of not having to wrap the sends and being more concise is probably worth it.

File: src/server/pfs/server/server_test.go [hunk 8]
```diff
@@ -2631,43 +2631,6 @@ func TestPutFileSplit(t *testing.T) {
...
-	require.NoError(t, c.FinishCommit(repo, commit.ID))
-
-	files, err = c.ListFile(repo, commit.ID, "line")
-	require.NoError(t, err)
-	require.Equal(t, 7, len(files))
-	for _, fileInfo := range files {
-		require.Equal(t, uint64(4), fileInfo.SizeBytes)
-	}
-}
-
```
Comment (jdoliner) [inline src/server/pfs/server/server_test.go]: Is this test no longer valid?

Comment (brycemcanally) [reply src/server/pfs/server/server_test.go]: This is testing the case where a user does a put file with a split then deletes files in the split. We discussed that this should not be allowed and instead the user should do a split using the atomic put file you are working on. This test breaks if we apply changes in mod revision order because of the way split file records are stored in etcd. The two deletes would get applied before the put file split in this case.

File: src/server/pkg/collection/iterator.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	Order    etcd.SortOrder
+	SelfSort bool
+}
+
+var DefaultOptions *Options = &Options{etcd.SortByCreateRevision, etcd.SortDescend, false}
+
+type etcdIterator interface {
+	next() ([]*mvccpb.KeyValue, error)
+	done() bool
+}
```
Comment (jdoliner) [inline src/server/pkg/collection/iterator.go]: One of the goals I had for this PR was to remove this iterator pattern from the code base, at least from the collections library and other parts that I'm touching in this PR. After using this pattern I'm convinced that it's not a very good way to do iteration in go for a couple of reasons:

1. It almost always leads to longer code, you can see this if you look above, you have code that reads:

```
iter := newEtcdIterator(c, prefix, limitPtr, opts)
for !iter.done() {
    kvs, err := iter.next()
    if err != nil {
	return err
    }
    // real code begins here
}
```

If this was using a callback it would look like:

```
return iterate(c, prefix,limitPtr, opts, func(kvs *mvccpb.KeyValues) error {
    // real code begins here
})
```

2 lines of boilerplate vs 8, this stuff really starts to add up too because anytime you want to compose one iterator to make another you need to write another for-loop. I probably cleaned up several hundred lines of iteration boilerplate as part of this PR.

2. There's a lot more reasonable ways to do iterator interfaces and as a result you have to read each individual iterator to see how it was implemented. In particular: some interfaces opt give you method to figure out if the iteration is complete, you've called yours done and it's true if iteration has finished, but it often makes more sense to have one that's true if iteration has not finished It's also common to have neither method but to have your `next` method return `ok` as it second return value if you can continue. Or sometimes it just returns an error and if you're at the end of the iteration that error is `io.EOF`. With a callback there's only one reasonable way to implement it.

3. There are certain optimizations that iterators make much harder. In particular, a really nice optimization for this code would be to asynchronously request the next batch will the first batch is being processed. To do that with iterators you would need to have a goroutine running in the background that you launch from you `next`, but now that goroutine could potentially leak, so you need to add a cancel or close method to your iterator, and callers of the iterator need to remember to defer that method everytime they create an iterator. Which is more boilerplate and potential bugs.

So long story short, let's migrate this over to callback interface like the other iteration methods in this PR and I'm going to add these thoughts on iteration to our official coding guidelines. I was actually planning to do that anyways as part of this PR.

**Comment by jdoliner** [comment]
One fairly large change requested, I can't submit and official review on this though because it's technically my PR.

File: src/server/pkg/collection/iterator.go (hunk 9 above)
Comment (brycemcanally) [reply src/server/pkg/collection/iterator.go]: This makes a lot of sense. I will go ahead and start migrating this over to a callback interface and make a mental note for when I add/change things in the future.

File: src/server/pkg/collection/iterator.go [hunk 10]
```diff
@@ -22,84 +23,71 @@ type Options struct {
...
-	c           *readonlyCollection
-	prefix      string
-	limitPtr    *int64
-	opts        *Options
-	compareFunc func(*mvccpb.KeyValue, *mvccpb.KeyValue) int
-	last        bool
-}
-
-func newEtcdIterator(c *readonlyCollection, prefix string, limitPtr *int64, opts *Options) etcdIterator {
+func iterate(c *readonlyCollection, prefix string, limitPtr *int64, opts *Options, f func(*mvccpb.KeyValue) error) error {
```
Comment (jdoliner) [inline src/server/pkg/collection/iterator.go]: This is a bit pedantic but how would you feel about renaming the functions in this file from `iterator` to `list` to match a little more with the other ones? I also think this function in particular is close enough to `listF` that you could just rename it to `listF` it seems like the only difference is that the `readonlyCollection` is an argument rather than the receiver.

Also if you want you could drop the `F`s from the ends of these functions, I only named them that because I was gradually migrating from code that had a bunch of functions called `List` already.

File: src/server/pfs/server/server_test.go (hunk 8 above)
Comment (jdoliner) [reply src/server/pfs/server/server_test.go]: Makes sense, could we replace it with a test that makes sure that we error when people try to do `put-file`s on open commits?

**Comment by jdoliner** [comment]
Few more minor requests.

**Comment by jdoliner** [comment]
All LGTM, again, can't officially approve this since I opened it.
